File: backend/src/classflow/domain/entities/attendance.py
from dataclasses import dataclass, field

from classflow.domain.entities.course import Course
from classflow.domain.entities.group import Group
from classflow.domain.entities.organization import OrganizationMember
from classflow.domain.enums import AttendanceStatus


@dataclass
class Attendance:
    id: int = field(init=False)
    organization_id: int = field(init=False)
    lesson_id: int
    student_id: int
    status: AttendanceStatus


@dataclass
class CourseAttendanceStats:
    course: Course
    group: Group | None
    teacher: OrganizationMember | None
    conducted_lessons: int
    present_lessons: int
    rate: int

File: frontend/src/entities/attendance.ts
import type { LessonDetail } from './lesson.ts';
import type { OrganizationMemberDetail } from './organization.ts';
import type { Group } from './group.ts';
import type { Course } from './course.ts';

export type AttendanceStatus = 'PRESENT' | 'ABSENT' | 'EXCUSED';

interface BaseAttendance {
  id: string;
  status: AttendanceStatus;
}

export interface Attendance extends BaseAttendance {
  lessonId: string;
  studentId: string;
}

export interface AttendanceDetail extends BaseAttendance {
  lesson: LessonDetail;
  student: OrganizationMemberDetail;
}

export interface AttendanceStats {
  present: number;
  absent: number;
  excused: number;
  total: number;
}

export interface WeekAttendanceStats {
  week: Date;
  present: number;
}

export interface CourseAttendanceStats {
  course: Course;
  group: Group | null;
  teacher: OrganizationMemberDetail | null;
  conductedLessons: number;
  presentLessons: number;
  rate: number;
}
